src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -83,9 +83,439 @@ module clsm_ensupd_read_obs
+
+    time_index = hh_num/3 +1
+    
+    ! -------------------------------------
+    ! eliminate no-data-values and data that fail initial QC
+    
+    j = 0
+    
+    do i=1,N_data
+       if (logit) write (logunit, *) 'hh_num=', hh_num
+       if (logit) write (logunit, *) 'time_index =', time_index
+
+       if ( (MODIS_SCA(i)>0.)           .and.         &  ! any neg is nodata
+            (CI_Index(i)>20)            .and.         &  ! Ignore obs
+            (Snow_QA(i)<3)              .and.         &
+            (lon(i)>=lon_subtime(time_index)) .and.         &
+            (lon(i)<lon_subtime(time_index+1))              & 
+           ) then     
+          
+          j=j+1
+          
+          MODIS_SCA(j) = MODIS_SCA(i)
+          lon(j)      = lon(i)
+          lat(j)      = lat(i)
+       end if
+       
+    end do
+   
+   if (logit) write (logunit, *) 'assimilation time', HH
+   if (logit) write (logunit, *) 'lat = ', lat
+   if (logit) write (logunit, *) 'lon = ', lon
+   if (logit) write (logunit, *) 'MODIS_SCA=', MODIS_SCA
+    
+    N_data = j
+    
+    deallocate(CI_Index)
+    deallocate(Cloud_Index)
+    deallocate(Snow_QA)
+  end subroutine read_MODISsca_hdf  
+
   ! *****************************************************************
+
+  subroutine read_obs_MODISsca(                               &
+  work_path, date_time, dtstep_assim, N_catd, tile_coord,     &
+  tile_grid_d, N_tile_in_cell_ij, tile_num_in_cell_ij,        &
+  this_obs_param, found_obs, MODIS_obs, std_MODIS_obs)
+
+  implicit none
+
+  !inputs
+   character(200), intent(in) :: work_path
+
+   type(date_time_type), intent(in):: date_time
+
+   integer, intent(in):: dtstep_assim, N_catd
+
+   type(tile_coord_type), dimension(:), pointer:: tile_coord
+   type(grid_def_type), intent(in) :: tile_grid_d
+
+   integer, dimension(tile_grid_d%N_lon, tile_grid_d%N_lat), intent(in):: &
+   N_tile_in_cell_ij
+   integer, dimension(:,:,:), pointer:: tile_num_in_cell_ij
+   
+   type(obs_param_type), intent(in):: this_obs_param
   
+   ! output
+   real, intent(out), dimension(N_catd):: MODIS_obs
+   real, intent(out), dimension(N_catd):: std_MODIS_obs
+   logical,intent(out) :: found_obs
+   logical :: file_exists
+
+   !locals
+   character(2):: MM, HH, MI
+   character(4):: YYYY
+   character(3):: DDD ! Day of Year 
+
+   character(300):: tmpfname1
+    
+   integer, dimension(N_catd)  ::  tmp_tile_id
+
+   integer :: i, ind, N_tmp, N_files
+   integer :: hh_num, mi_num
+
+   integer:: dtstep_assim_max=10800
+
+   character(300), dimension(:), allocatable :: fnames
+   
+   real, dimension(:), pointer:: tmp_obs, tmp_lat, tmp_lon
+   integer, dimension(:), pointer:: tmp_tile_num
+   
+   integer, dimension(N_catd):: N_obs_in_tile
+
+   character(len=*), parameter :: Iam = 'read_obs_MODISsca'
+   character(len=400) :: err_msg
+
+   nullify (tmp_obs, tmp_lat, tmp_lon, tmp_tile_num)
+
+   write(HH, '(i2.2)') date_time%hour
+   read(HH, '(I10)') hh_num
```

Comment:
Why convert "date_time%hour" into a string and then back to an integer?  The modulus calculation below should work just fine as "mod(date_time%hour,3)".  Likewise for minutes and seconds, which should also be checked.